<a href="https://colab.research.google.com/github/JennaBeutler/APRD-6342/blob/master/Exercise_3_LASSO_regression_for_fat_sales_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Exercise 3: LASSO regression for fat sales data
### APRD 6342  
### Fall 2019  
### Jenna Beutler

Importing all necessary packages and mounting the drive  
uploading data into Colab

In [0]:
import pandas as pd
import pandas
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split 

from google.colab import drive
drive.mount('/content/drive')

In [0]:
path = '/content/drive/My Drive/Colab Notebooks/Python Exercises/finalmaster-ratios.csv'
alldata = pd.read_csv(path)
alldata.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

For the sake of this exercise, we know the first 8 variables aren't relevant to the regression, and we want to predict the Purchases column, so we will remove them from the list of variables in our regression

In [0]:
allvariablenames = list(alldata.columns.values)
allvariablenames.remove('# Purchases')
allvariablenames.remove('B01001001')
allvariablenames.remove('B01001002')
allvariablenames.remove('B01001003')
allvariablenames.remove('B01001004')
allvariablenames.remove('B01001005')
allvariablenames.remove('B01001006')
allvariablenames.remove('B01001007')


Setting up the predictors, the targets, and the train, test split then running and fitting our LASSO model

In [0]:
predictors = alldata[allvariablenames]  
target = alldata['# Purchases']

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123) 

In [0]:
model = LassoLarsCV(cv=10, precompute=False)
model.fit(predictors, target)

In [0]:
# turning the model predictions/ coefficients into a dataframe, which is easier to work with 
predictors_model = pd.DataFrame(predictors.columns) 

# naming the first column 'label' which is the name of each predictor variable
predictors_model.columns = ['label']

# creating the second column of coefficients and pulling the coefficients from the model to fill in the dataframe
predictors_model['coeff'] = model.coef_

#prints the new dataframe
predictors_model

label     coeff
0    B01001008  0.000000
1    B01001009  0.000000
2    B01001010  0.000000
3    B01001011  0.000000
4    B01001012  0.000000
..         ...       ...
177  B19001013 -0.002185
178  B19001014  0.000000
179  B19001015  0.000000
180  B19001016  0.000000
181  B19001017  1.886081

[182 rows x 2 columns]

Looking at what coefficients from the LASSO model are greater than 0

In [0]:
# loops through each row of the above dataframe
for index, row in predictors_model.iterrows():
    if row['coeff'] > 0: # if the coefficient is greater than 0, print the row.
        print(row.values)
# we want to look at all coefficients greater than 0 since they are all significant under a LASSO model. 

['B01001014' 0.5330720656233966]
['B01001036' 1.6591227573088478]
['B01001037' 2.4482060101637564]
['B01001038' 2.21381020716577]
['B02001005' 0.9859155344708008]
['B02001007' 0.006591090821753925]
['B13014016' 0.0061795097515955225]
['B13014026' 0.574990933842321]
['B13014027' 1.2849909864550721]
['B13016008' 0.31692716289106393]
['B15002027' 3.9418447021477347]
['B19001017' 1.8860806124452758]


**QUESTION #1: In your own words, explain what the above lines of code are doing. Why am I doing it? Explain each line.**   
(commented in code)


**Now we know which variables most predicted sales. Because of the aforementioned feature section feature of LASSO, any coefficients that are non-zero are significant, and thus variables that truly predict unique amounts of sales. The higher the value of the coefficient, the steeper the relationship between sales and that variable.**

**Let's figure out what these Census variables actually mean. To do this, we're going to perform a simple hack, and have you Google the variable name + "Census", here's an example. Click on the first result from socialexplorer.com. For this example, I'd report: Females aged 30 to 34 Years.**  
B01001014: Males 40 - 44  
B01001036: females 30 - 34  
B01001037: females 35 - 39  
B01001038: females 40 - 44   
B02001005: Asian  
B02001007: Some other Race  
B13014016: now married  
B13014026: women 15-50 that had a baby in the past year   
B13014027: women 15-50, unmarried, no baby in past year with graduate degree  
B13016008: women 40-44 that had a baby in past year  
B15002027: 25+ females 12th grade no diploma  
B19001017: Households that make 200,000+


**QUESTION #2: Interpret each variable intuitively. What Census variables most predict sales? What does that mean, practically? Here's what I'd put for the example above. "In areas where there are more females aged 30-34, we sell more Bobo Bars."**  
Men and women roughly between the ages of 30 and 44 are more likely to buy bobo bars (fairly strong).  
Non-white, especially those that fall into the Asian and Other categories of race are slightly more likely to buy bobo bars (very small relationship).  
Now married women are also slightly more likely to buy bobo bars (very slight relationship).  
Areas with Women that had a baby in the past year will sell more bars (medium strength relationship). 
Unmarried women with graduate degrees are more likely to buy bars(medium strength relationship).  
Areas with women 25+ and no diploma will likely sell more bars(VERY strong).  
wealthier areas will probably sell more bars as well (strong).  

**QUESTION #3: If I had to report only two census variables to my boss that most steeply predicted sales, what would those be?**  
The 25+ women with no high school diploma (3.9 coef)  
females 35 - 39 (2.45 coef) 



Next, lets take a look at the mean squared error for the training and training set:

In [0]:
import sklearn as sk
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import LassoCV, LarsCV, LassoLarsCV
from sklearn.metrics import mean_squared_error, r2_score, median_absolute_error, accuracy_score

In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
21060.761034858362


In [0]:
train_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('test data MSE')
print(train_error)

test data MSE
36561.48369809522


**QUESTION #4: Are the training and text set mean squared errors similar? What does that mean practically?**  
The standard errors are both quite large, meaning that the model isn't super accurate in predicting sales of bobo bars to begin with. The goal of machine learning / regressions is to minimize the mean squared error, and this test did not. Practically, this means that this particular model isn't exactly the best for predicting who buys bobo bars. 

In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2733096196457573


In [0]:
rsquared_test=model.score(pred_test,tar_test)
print ('test data R-square')
print(rsquared_test)

test data R-square
0.27480133108928484


**QUESTION #5: If your boss asked, "How well does Census data, overall, predict sales?" What would you say? Why?**  
I would say that census data doesn't really predict sales because of the high mean squared errors and fairly low R-square scores. The R-squares mean that we can only predict/explain about 27 percent of the data. 

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
-3.9606645084610435


**QUESTION #6: What is our baseline sales number? What does that mean, practically? Think back to what y-intercepts mean in regression models.**  
The y intercept is -3.96 which means that if all other variables equaled 0 (which isn't practically possible) that sales would be in the negatives. However, this just helps us scale the number of sales that actually happen given the variables that we use. 